tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc

```diff
@@ -30,6 +31,63 @@ namespace mlir {
 namespace disc_ral {
 namespace {
 
+// Returns true if a tensor is constant and all its elements are equal to
+// `elemVal`
+bool allElementsAreSameValue(Value tensor, int64_t elemVal) {
+  auto definingOp = tensor.getDefiningOp();
+  if (!definingOp) return false;
+  if (isa<mhlo::BroadcastOp, mhlo::BroadcastInDimOp,
+          mhlo::DynamicBroadcastInDimOp>(definingOp)) {
+    return allElementsAreSameValue(definingOp->getOperand(0), elemVal);
+  }
+  DenseElementsAttr denseAttr;
+  if (!matchPattern(tensor, m_Constant(&denseAttr))) return false;
+  if (denseAttr.getNumElements() != 1 && !denseAttr.isSplat()) return false;
+
+  Type elemTy = denseAttr.getElementType();
+  if (elemTy.isIntOrIndex()) {
+    return (*denseAttr.getValues<APInt>().begin()).getSExtValue() == elemVal;
+  } else if (elemTy.isa<FloatType>()) {
+    return (*denseAttr.getValues<APFloat>().begin()).convertToDouble() ==
```

Comment:
nit: why convert to double for comparison with an int64_t value?

Reply:
1, convert f16/f32/f64 to a unified type f64. 
2, We only support the case where `elemVal` is int64_t (as indicated by its type), thus it does not matter: c++ implicit conversion will cast to f64.
3, If we need to support `elemVal`  to double, we need to overload this function.

Reply:
A template is simple enough? Also feels weird to compare double and int.